TheJumpCloud/jcapi-powershell
# SA-2318 Automate Changelog on SDK Release

## Issues
* [SA-2318](https://jumpcloud.atlassian.net/browse/SA-2318) - Automate Changelog on SDK release

## What does this solve?

There should be a changelog when we release new versions of our SDKs. This PR addresses that. 
In addition a process is introduced to the CI workflow to automatically update the git diff of the files changed in the sdk's /generated/custom directory. 

Lastly on merge to main branch the CI process will automatically draft a release of each SDK and add the corresponding version notes from the changelog file. Each changelog file is located in the root of the repo ex. JumpCloud.SDK.V1.md

## Is there anything particularly tricky?

This is PR is not intended to result in a SDK release. 

## How should this be tested?

The Build-SDKChangelog function can be used to test the functionality of this PR. 

Test 1)
Update the PSD1 version of an SDK function & from the root of the project run `Tools/BuildSDKChangelog -SDKName JumpCloud.SDK.V1` (or which ever SDK you changed).
The ChangeLog File for the corresponding SDK should be updated with a new version block containing the version you've specified in the PSD1 File.

Ex. Modify /SDKs/PowerShell/JumpCloud.SDK.V1/JumpCloud.SDK.V1.psd1 module version to 2.0.0
Run `Tools/BuildSDKChangelog -SDKName JumpCloud.SDK.V1` from the root of the project
'./JumpCloud.SDK.V1.md' should have been updated with a 2.0.0 version header

Test 2)
After running test 1, fill out some data in the {{ Replace me blocks }} and run `Build-SDKChangelog -SDKName JumpCloud.SDK.V1` again, nothing should have changed in the changelog file.

Test 3)
After running test 2:
locally add any .ps1 file to /SDKs/PowerShell/JumpCloud.SDK.V1/custom/generated
locally modify any existing .ps1 file to /SDKs/PowerShell/JumpCloud.SDK.V1/custom/generated
locally remove any existing .ps1 file in /SDKs/PowerShell/JumpCloud.SDK.V1/custom/generated

locally commit changes but **do not** push to remote.

Run `./Tools/Build-SDKChangelog -SDKName JumpCloud.SDK.V1` The autogenerated section should update and display your added, modified & removed file in the "Functions Added, Modified & Removed" blocks

## Screenshots
Changelog File:
<img width="1238" alt="Screen Shot 2022-03-25 at 8 22 25 AM" src="https://user-images.githubusercontent.com/54448601/160139528-40018fde-e940-46f7-a973-c3b5bcc02885.png">
On SDK merge to main, a draft release is autogenerated as a CI task:
<img width="1282" alt="Screen Shot 2022-03-25 at 8 25 29 AM" src="https://user-images.githubusercontent.com/54448601/160139774-6848eec0-b07e-40d1-a9d6-f090142a1c75.png">
